flutter/lib/ui/geometry.dart

```diff
@@ -769,48 +886,62 @@ class RRect {
   Rect get outerRect => new Rect.fromLTRB(left, top, right, bottom);
 
   /// The non-rounded rectangle that fits inside this rounded rectangle by
-  /// touching the middle of each curved corner.
+  /// touching the middle of the inner-most curved corners.
   Rect get safeInnerRect {
     const double kInsetFactor = 0.29289321881; // 1-cos(pi/4)
+
+    final double leftRadius = math.max(blRadiusX, tlRadiusX);
+    final double topRadius = math.max(tlRadiusY, trRadiusY);
+    final double rightRadius = math.max(trRadiusX, brRadiusX);
+    final double bottomRadius = math.max(brRadiusY, blRadiusY);
+
     return new Rect.fromLTRB(
-      left + radiusX * kInsetFactor,
-      top + radiusY * kInsetFactor,
-      right - radiusX * kInsetFactor,
-      bottom - radiusY * kInsetFactor
+      left + leftRadius * kInsetFactor,
+      top + topRadius * kInsetFactor,
+      right - rightRadius * kInsetFactor,
+      bottom - bottomRadius * kInsetFactor
     );
   }
 
-  /// The rectangle that would be formed using only the straight sides of the
-  /// rounded rectangle, i.e., the rectangle formed from the centers of the
-  /// ellipses that form the corners. This is the intersection of the
-  /// [wideMiddleRect] and the [tallMiddleRect].
+  /// The rectangle that would be formed using the axis-aligned intersecion of
+  /// the slides of the rectangle, i.e., the rectangle formed from the
+  /// inner-most centers of the ellipses that form the corners. This is the
+  /// intersection of the [wideMiddleRect] and the [tallMiddleRect].
```

Comment:
this could end up having negative dimensions, right? If so, mention that in the description and give an example